src/dds_data.py

```diff
@@ -10,208 +10,333 @@
 import logging
+import time
+import copy
+from queue import Queue
+from typing import Dict, List
+import gc
 
 from dds_service import builtin_observer
-from dds_qos import qos_match
+from dds_qos import qos_match, dds_qos_policy_id
 from utils import singleton, EntityType
 
 
+class DataEndpoint:
+    def __init__(self, endpoint: DcpsEndpoint, entity_type) -> None:
+        self.endpoint: DcpsEndpoint = endpoint
+        self.entity_type: EntityType = entity_type
+        self.participant = None
+        self.missmatches : Dict[str, List[dds_qos_policy_id]]= {}
+
+    def isReader(self):
+        return self.entity_type == EntityType.READER
+
+    def isWriter(self):
+        return self.entity_type == EntityType.WRITER
+
+    def link_participant(self, participant: DcpsParticipant):
+        if str(self.endpoint.participant_key) == str(participant.key):
+            self.participant = participant
+
+
+class DataTopic:
+    def __init__(self, name) -> None:
+        self.name = name
+        self.reader_endpoints: Dict[str, DataEndpoint] = {}
+        self.writer_endpoints: Dict[str, DataEndpoint] = {}
+
+    def add_endpoint(self, endpoint: DataEndpoint):
+        if endpoint.isReader():
+            if str(endpoint.endpoint.key) not in self.reader_endpoints:
+                self.reader_endpoints[str(endpoint.endpoint.key)] = endpoint
+        else:
+            if str(endpoint.endpoint.key) not in self.writer_endpoints:
+                self.writer_endpoints[str(endpoint.endpoint.key)] = endpoint
+
+        self.check_qos_mismatch(endpoint)
+
+    def remove_endpoint(self, endpointKey: str):
+        if endpointKey in self.reader_endpoints:
+            for mimKey in self.reader_endpoints[endpointKey].missmatches:
+                if mimKey in self.writer_endpoints:
+                    if endpointKey in self.writer_endpoints[mimKey].missmatches:
+                        del self.writer_endpoints[mimKey].missmatches[endpointKey]
+
+            del self.reader_endpoints[endpointKey]
+    
+        if endpointKey in self.writer_endpoints:
+            for mimKey in self.writer_endpoints[endpointKey].missmatches:
+                if mimKey in self.reader_endpoints:
+                    if endpointKey in self.reader_endpoints[mimKey].missmatches:
+                        del self.reader_endpoints[mimKey].missmatches[endpointKey]
+
+            del self.writer_endpoints[endpointKey]
+
+    def link_participant(self, participant: DcpsParticipant):
+        for reader_key in self.reader_endpoints.keys():
+            self.reader_endpoints[reader_key].link_participant(participant)
+        for writer_key in self.writer_endpoints.keys():
+            self.writer_endpoints[writer_key].link_participant(participant)
+
+    def hasEndpoints(self) -> bool:
+        return len(self.reader_endpoints) > 0 or len(self.writer_endpoints) > 0
+
+    def check_qos_mismatch(self, data_endpoint: DataEndpoint):
+
+        endpoints_to_check = self.reader_endpoints
```

Comment:
so you mean it could be totally fine to same topic name `x` with readers and writers of different types?
From my experience with DDS systems that sounds like not a good idea, but of course when dds allows it we should look at it.
I will add an issue for that to look at it later.